torchdynamo/guards.py

```diff
@@ -210,32 +250,45 @@ def EQUALS_MATCH(self, guard: Guard):
                 np.uint32,
                 np.uint64,
             ),
-        ), type(val).__name__
+        ), t.__name__
         if istype(val, (torch.device, torch.dtype)):
             # TODO(jansel): is this slow? perhaps optimize it
-            self.code.append(f"str({ref}) == {str(val)!r}")
+            code = f"str({ref}) == {str(val)!r}"
+            self.code.append(code)
+            guard.set_export_info("EQUALS_MATCH", t, code, obj_id)
             return
 
         # Special case for nan because float("nan") == float("nan") evaluates to False
         if istype(val, float) and math.isnan(val):
-            self.code.append(f"___check_type_id({ref}, {self.id_ref(type(val))})")
-            self.code.append(f"__math_isnan({ref})")
+            code = f"___check_type_id({ref}, {self.id_ref(t)})"
+            export_code = code
+            self.code.append(code)
+            code = f"__math_isnan({ref})"
+            self.code.append(code)
+            export_code = f"{export_code} and {code}"
+            guard.set_export_info("EQUALS_MATCH", t, export_code, obj_id)
```

Comment:
at least for now this seems fine, but i wonder if it makes sense to accumulate 'guard objects' which have their guard expression built in into self, rather than putting the guard code into self.code separately from the guard object itself.

Reply:
Yeah, I think a very reasonable refactor is to stop passing code around as str, and instead make guard objects. This was a sort of path-of-least-resistance for a proposal in today's code.